[CodeGen][SDAG] Remove Combiner WorklistMap

aengelke opened the pull request:
DenseMap for pointer lookup is expensive, and this is only used for deduplication and index lookup. Instead, store the worklist index in the node itself.

This brings a substantial performance improvement, see [here (compile-time-tracker)](http://llvm-compile-time-tracker.com/compare.php?from=9d70975c7a72f3fa58d2d63090b92886dbf8a32b&to=13eff8c73d11595da9c01a0bb6580fea698aa3c9&stat=instructions:u).

aengelke committed:
[CodeGen][SDAG] Remove Combiner WorklistMap

DenseMap for pointer lookup is expensive, and this is only used for
deduplication and index lookup. Instead, store the worklist index in the
node itself.

jayfoad commented on llvm/include/llvm/CodeGen/SelectionDAGNodes.h:
```diff
@@ -619,6 +619,9 @@ END_TWO_BYTE_PACK()
   int NodeId = -1;
 
+  /// Index in worklist of DAGCombiner, or -1.
+  int CombinerWorklistIndex = -1;
```
Does this increase the size of the node, or does it fill a hole (when pointers are 64-bit)?

aengelke replied:
This fills 4 padding bytes, the size of an SDNode remains unchanged at 88 bytes (x86-64).

jayfoad reviewed (commented):
> This brings a substantial performance improvement, see [here (compile-time-tracker)](http://llvm-compile-time-tracker.com/compare.php?from=9d70975c7a72f3fa58d2d63090b92886dbf8a32b&to=13eff8c73d11595da9c01a0bb6580fea698aa3c9&stat=instructions:u).

Nice!

RKSimon commented:
Love this!

Although I think it takes all the padding bits we'd be considering using to extend NumOperands/NumValues to U32 to fix #55737 etc. :(

arsenm reviewed (commented):
Can we do something with PersistentId instead? Or at least, steal the bytes in release builds?

aengelke commented:
> Can we do something with PersistentId instead? Or at least, steal the bytes in release builds?

I guess, we probably could? I'm not too familiar how SDAG behaves under pressure, but I can imagine that the worklist could exceed 64k nodes, so the worklist index would need 32 bits. SDNodeFlags is 2 bytes (14 bits), so there are 2 padding bytes near the end before CFIType (which btw appears to be only used for KCFI calls). When swapping Flags and PersistentId, we should get 32 bits where to put the combiner index (+ place the PersistentId under ENABLE_ABI_BREAKING_CHECKS, see [D120714](https://reviews.llvm.org/D120714)). (NB: just from glancing at the code, I haven't compile-checked.)

> it takes all the padding bits we'd be considering using to extend NumOperands/NumValues to U32 to fix https://github.com/llvm/llvm-project/issues/55737 etc. :(

Well... I *personally* think that nobody should use large aggregates as IR values (especially not with >65k values) in the first place -- it's been discouraged for as long as I can remember -- and that this is something we shouldn't need to support. It's probably easier (and also in their interest w.r.t. compile times) to generate different IR in the front-end.

aengelke committed:
Shuffle SDNode fields

This feels a bit pointless at this stage, but keeps the four padding
bytes in the middle of the struct for future use.

aengelke commented:
I moved the CombinerWorklistIndex to the end, so there are currently two padding bytes in the middle for future use. Two further bytes, e.g. for expanding NumValues/NumOperands to 32 bits each, could be gained in future by placing PersistentId under LLVM_ENABLE_ABI_BREAKING_CHECKS.

aengelke committed:
Fix visibility of fields

+ another reorder to avoid unnecessary visibility changes

RKSimon reviewed (approved):
LGTM - cheers!
